sage/square_root_banderwagon_and_bandersnatch.sage

```diff
@@ -0,0 +1,33 @@
+#Parameters
+p = 52435875175126190479447740508185965837690552500527637822603658699938581184513
+Fp = GF(p)
+
+# BaseField2Adicity = 32  #see https://github.com/crate-crypto/go-ipa/blob/408dbffb2041271c95979a3fb79d98b268bf2880/bandersnatch/fp/sqrt.go#L22
+# sqrtPrecomp_PrimitiveDyadicRoots[] is an array of size BaseField2Adicity + 1.
+# sqrtPrecomp_PrimitiveDyadicRoots[0] should be equal to 10238227357739495823651030575849232062558860180284477541189508159991286009131 (see https://github.com/crate-crypto/go-ipa/blob/408dbffb2041271c95979a3fb79d98b268bf2880/bandersnatch/fp/sqrt.go#L46)
+sqrtPrecomp_PrimitiveDyadicRoots = {}
+sqrtPrecomp_PrimitiveDyadicRoots[0] = 10238227357739495823651030575849232062558860180284477541189508159991286009131
+
+print('p  : ' + p.hex())
+
+print('\n\nPrimitive Dyadic Roots:\n')
+# function sqrtPrecomp_PrimitiveDyadicRoots:
+print(hex(sqrtPrecomp_PrimitiveDyadicRoots[0]))
+for i in range(0, 32):
+    sqrtPrecomp_PrimitiveDyadicRoots[i+1] = Fp(sqrtPrecomp_PrimitiveDyadicRoots[i]^2)
+    print(hex(sqrtPrecomp_PrimitiveDyadicRoots[i+1]))
+
+
+# function sqrtPrecomp_PrecomputedBlocks:
+block = {}
+print('\n\nPrecomputed Blocks:\n')
+for i in range (0, 4):
+    block[i] = {}
+    block[i][0] = 1
+    for j in range (1, 256):
+        block[i][j] = Fp(block[i][j-1] * sqrtPrecomp_PrimitiveDyadicRoots[i * 8])
+        print(hex(block[i][j]))
+
+# function sqrtPrecomp_dlogLUT:
```

Comment:
For the implementations you need to convert a point into limbs
Limbs are of 64bits, you can achieve this by 

Assuming that this is related to the Jubjub Edwards curve, for which `d = -(10240/10241) (mod r = 52435875175126190479447740508185965837690552500527637822603658699938581184513) = 19257038036680949359750312669786877991949435402254120286184196891950884077233`, the four 64-bit limbs (in little-endian order) are:

`(73851820219580081, 2967167836563676454, 17725484271033745364, 3067823152860638024)`

The "process" is pretty simple:

- `73851820219580081 = d mod 2^64`
- `2967167836563676454 = (d / 2^64) mod 2^64`
- `17725484271033745364 = (d / 2^128) mod 2^64`
- `3067823152860638024 = (d / 2^192) mod 2^64`

Hope this will help in your implementation, as limbs are not there in sage by default AFAIK

Reply:
Just using the field printer as hex is fine, Constantine can convert hex to fields at compile-time and it's easier to debug.

Reply:
Yes
But this was for generating the LUT

Check this line :  https://github.com/crate-crypto/go-ipa/blob/408dbffb2041271c95979a3fb79d98b268bf2880/bandersnatch/fp/sqrt.go#L80